Transfer data Rust->Dart as TransferableTypedData

**Is your feature request related to a problem? Please describe.**
In our app we are using flutter_rust_bridge to read data from a file, decrypt and decompress it. The result ends up in Flutter as `Uint8List`. Later we pass this data from the main isolate to a separate isolate. Sending the data (200 - 300 MiB) to the isolate takes  around 100 ms which causes the UI to stutter. I learned that I can convert the `Uint8List` to a `TransferableTypedData` type (like this: `TransferableTypedData.fromList([uint8List])`). After that, the transfer takes significantly less than 1 ms. However now the UI lags due to `fromList` also taking 50 - 100 ms.
For unrelated reasons I can only read the data / interface with flutter_rust_bridge on the main isolate.

**Describe the solution you'd like**
In some cases the main isolate itself will never access the data and will pass it directly on to the other isolate. In those cases it would be very convenient to already get the data as `TransferableTypedData` from flutter_rust_bridge. In case we need access to the data on the main isolate we can always call `materialize()`.

**Describe alternatives you've considered**
Use flutter_rust_bridge from the other isolate directly.

**Additional context**
Add any other context or screenshots about the feature request here.

Thank you for the quick feedback!

I agree, that the process of reading the data should happen directly on the second isolate. Currently my app can't do that because the wrapper type for "EncryptedFile" contains fields which can't be sent over the isolate boundary. This is fixable but would take some large refactoring.

Exploring the other solution using private data inside a struct: I have set up a new project and tried to create a struct with a private field like this:

```rust
pub struct MyData {
    private_data: Vec<u8>,
}

impl MyData {
    pub fn new(number_of_bytes: usize) -> MyData {
        let data = vec![42u8; number_of_bytes];
        MyData { private_data: data }
    }

    pub fn materialize(self) -> Vec<u8> {
        self.private_data
    }
}
```

and ran `flutter_rust_bridge_codegen generate` (version 2.7.1).
However this seems to generate invalid rust code:
```
error[E0616]: field `private_data` of struct `MyData` is private
   --> src/frb_generated.rs:277:15
    |
277 | ...   [self.private_data.into_into_dart().into_dart()].into_...
    |             ^^^^^^^^^^^^ private field

error[E0616]: field `private_data` of struct `MyData` is private
   --> src/frb_generated.rs:307:36
    |
307 |         <Vec<u8>>::sse_encode(self.private_data, serializer);
    |             

```

Do I have to annotate something manually?

--- a/website/docs/guides/how-to/ndk-init.md
+++ b/website/docs/guides/how-to/ndk-init.md
@@ -100,8 +100,8 @@ class MainActivity : FlutterActivity() {
     override fun configureFlutterEngine(
         @NonNull flutterEngine: FlutterEngine,
     ) {
-        super.configureFlutterEngine(flutterEngine)
         flutterEngine.plugins.add(MyPlugin())
+        super.configureFlutterEngine(flutterEngine)
     }
 }
 
@@ -147,18 +147,26 @@ jni = "0.21"
 
 ```rust
 #[cfg(target_os = "android")]
-use {
-    jni::{objects::JClass, objects::JObject, JNIEnv},
-    mylib::setup_android,
-};
+mod init_android_context {
+    use jni::{objects::JClass, objects::JObject, objects::GlobalRef, JNIEnv};
+    use std::sync::OnceLock;
+    use std::ffi::c_void;
 
-#[cfg(target_os = "android")]
-#[no_mangle]
-pub extern "system" fn Java_com_example_spareshare_MyPlugin_init_1android(
-    env: JNIEnv,
-    _class: JClass,
-    ctx: JObject,
-) {
-    setup_android(env, ctx);
+    static CTX: OnceLock<GlobalRef> = OnceLock::new();
+
+    #[no_mangle]
+    pub extern "system" fn Java_com_example_local_1auth_MyPlugin_init_1android(
+        env: JNIEnv,
+        _class: JClass,
+        ctx: JObject,
+    ) {
+        let global_ref = env.new_global_ref(&ctx).expect("to make global reference");
+        let vm = env.get_java_vm().unwrap();
+        let vm = vm.get_java_vm_pointer() as *mut c_void;
+        unsafe {
+            ndk_context::initialize_android_context(vm, global_ref.as_obj().as_raw() as _);
+        }
+        CTX.get_or_init(|| global_ref);
+    }
 }
 ```

--- a/website/docs/guides/how-to/ndk-init.md
+++ b/website/docs/guides/how-to/ndk-init.md
@@ -1,11 +1,14 @@
 # Android NDK Init
 
+:::info
 This page is only needed to read if you want to use Android NDK in Rust code,
 and see "android context was not initialized" error, or want to know more about`ndk_context` initialization.
+:::
 
 Related issues:
 [#1323](https://github.com/fzyzcjy/flutter_rust_bridge/issues/1323),
-[#1868](https://github.com/fzyzcjy/flutter_rust_bridge/issues/1868).
+[#1868](https://github.com/fzyzcjy/flutter_rust_bridge/issues/1868),
+[#2466](https://github.com/fzyzcjy/flutter_rust_bridge/issues/2466).
 
 ## Method 1
 

--- a/website/docs/guides/how-to/transferable-typed-data.md
+++ b/website/docs/guides/how-to/transferable-typed-data.md
@@ -1,0 +1,3 @@
+# TransferableTypedData
+
+TODO

--- a/website/docs/guides/how-to/transferable-typed-data.md
+++ b/website/docs/guides/how-to/transferable-typed-data.md
@@ -1,3 +1,5 @@
 # TransferableTypedData
 
-TODO
+If you want to use Dart's `TransferableTypedData` to avoid data copying between main isolate and a second isolate,
+then please refer to the issue https://github.com/fzyzcjy/flutter_rust_bridge/issues/2520 about an alternative approach
+that also avoids data copying.

--- a/website/docs/guides/miscellaneous/compatibility.md
+++ b/website/docs/guides/miscellaneous/compatibility.md
@@ -3,6 +3,11 @@
 Since flutter_rust_bridge v2, we will try our best to follow [semantics versioning](https://semver.org/)
 with the following exceptions.
 
+:::info
+flutter_rust_bridge will automatically upgrade the versions as well as checking them.
+So usually there is no need to care about things below manually.
+:::
+
 Firstly, all flutter_rust_bridge-related packages will need to have exactly the same version. They are:
 
 * Dart package `flutter_rust_bridge` (runtime support)
@@ -16,8 +21,6 @@ In addition, it seems better to put the limited resources to more important thin
 [As is pointed out](https://github.com/fzyzcjy/flutter_rust_bridge/issues/1432#issuecomment-1849199332),
 this is not very uncommon, e.g. Apple's Safari is kept in sync with MacOS version,
 and this should not be a problem with upgrade tools.
-
-(WIP: (1) Make a tool to check these versions agree at runtime or as a doctor command (2) tool to upgrade them in one command.)
 
 Secondly, some classes with comments explicitly mentioning breaking-change things,
 especially the `Handler` classes, may contain breaking change in minor version bump.

--- a/website/docs/guides/types/arbitrary/rust-auto-opaque/override-defaults.md
+++ b/website/docs/guides/types/arbitrary/rust-auto-opaque/override-defaults.md
@@ -3,6 +3,9 @@
 By default, flutter_rust_bridge tries to infer whether a type is opaque or not automatically.
 To override the default behavior, please use `#[frb(opaque)]` and `#[frb(non_opaque)]`.
 
+Sometimes, the opaqueness cannot be overrided. For example, when struct `A` is used in a function as `&mut A`.
+(This is because non-opaque types are copied field-by-field, thus it is meaningless to have a mutable reference.)
+
 ## Example
 
 If a type is indeed encodable, it will by default be translated to the corresponding Dart types.

--- a/website/docs/manual/troubleshooting.md
+++ b/website/docs/manual/troubleshooting.md
@@ -145,6 +145,14 @@ Check [the related issue on GitHub](https://github.com/fzyzcjy/flutter_rust_brid
 
 Please refer to [#2425](https://github.com/fzyzcjy/flutter_rust_bridge/issues/2425) for more details.
 
+Shortly speaking, it seems flutter_rust_bridge cannot do much, just like the same scenario in [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz/issues/372).
+Instead, users may need to add the section below in `Cargo.toml`:
+
+```toml
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(frb_expand)'] }
+```
+
 ## Other problems?
 
 Don't hesitate to [open an issue](https://github.com/fzyzcjy/flutter_rust_bridge/issues/new/choose)! I usually reply

--- a/website/docs/manual/troubleshooting.md
+++ b/website/docs/manual/troubleshooting.md
@@ -148,7 +148,7 @@ Please refer to [#2425](https://github.com/fzyzcjy/flutter_rust_bridge/issues/24
 Shortly speaking, it seems flutter_rust_bridge cannot do much, just like the same scenario in [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz/issues/372).
 Instead, users may need to add the section below in `Cargo.toml`:
 
-```toml
+```
 [lints.rust]
 unexpected_cfgs = { level = "warn", check-cfg = ['cfg(frb_expand)'] }
 ```

--- a/website/sidebars.js
+++ b/website/sidebars.js
@@ -413,6 +413,7 @@ const sidebars = {
                         'guides/how-to/rust-compilation',
                         'guides/how-to/cargo-workspaces',
                         'guides/how-to/cross-origin',
+                        'guides/how-to/transferable-typed-data',
                     ],
                 },
                 'guides/users',

